model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProvider.java

```diff
@@ -607,6 +722,11 @@ public void removeUserSessions(RealmModel realm) {
         clusterEventsSenderTx.addEvent(
                 RemoveUserSessionsEvent.createEvent(RemoveUserSessionsEvent.class, InfinispanUserSessionProviderFactory.REMOVE_USER_SESSIONS_EVENT, session, realm.getId(), true),
                 ClusterProvider.DCNotify.LOCAL_DC_ONLY);
+
+        log.debugf("Removing user sessions for realm via persistence provider. realm=%s", realm.getId());
```

Comment:
`RealmAdminResource#logoutAll` called when admin-console -> "logout all sessions" is invoked.

The infinispan code path invalidates and clears all sessions from memory, hence offline-tokens can no longer be used.
However, the offline user sessions remain listed in the database, but they cannot be reclaimed / reused anymore. Even a sever restart doesn't bring them life.

I think this dangling (offline) user-sessions needs to be cleared from the database, if the sessions are removed.

Reply:
It seems to me that `RealmAdminResource#logoutAll` doesn't logout offline sessions in current master? I see this based on some testing and also when looking at the code. The `REMOVE_USER_SESSIONS_EVENT` registered in the InfinispanUserSessionsProviderFactory calls the `provider.onRemoveUserSessionsEvent(sessionEvent.getRealmId())` and that one seems to remove just "online" sessions?

The behaviour of "Logout all" in the admin console is not very clear. It is not possible to differentiate between "Logout all online and offline" sessions or "Logout just all online sessions" etc. I think we have some JIRAs to improve that. But my vote is, that this PR focus just on the offline sessions preloading and doesn't change any other existing behaviour and hence keep "Logout all" to still focus just on online sessions.

Reply:
Okay, thanks for clarification. I'll remove the offline-session removal from the "logout all" call.

Reply:
May I leave the code `UserSessionPersisterProvider#removeUserSessions(realm,offline)` in? 
This could then at least be used by users in custom extensions who need this functionality.
Btw. I know of several customers who need a true logout-all feature including offline user-sessions.